NeoML/src/TraditionalML/SMOptimizer.cpp

```diff
@@ -241,202 +104,383 @@ void CKernelCache::lruInsert(CList *l)
+		updateAlphaStatusAndGradient0( i );
+		updateAlphaStatusAndGradient0( j );
 	}
-	if(log != 0) {
+	
+	// Calculate the free term
+	freeTerm = calculateFreeTerm();
+	if(log != nullptr) {
 		*log << "\noptimization finished, #iter = " << t << "\n";
+		*log << "free term = " << freeTerm << "\n";
+	}
+
+	if( shrinking ) {
+		// put back the solution
+		_alpha.SetSize( l );
+		for( int i = 0; i < l; ++i ) {
+			_alpha[activeSet[i]] = alpha[i];
+		}
 	}
-	// Calculate the free term
-	freeTerm = calculateFreeTerm( alpha, gradient );
 }
 
-// return i,j such that
+// return 1 if already optimal, return 0 otherwise
 // i: maximizes -y_i * grad(f)_i, i in I_up(\alpha)
-// j: minimizes -y_i * grad(f)_i, i in I_low(\alpha)
-void CSMOptimizer::findMaximalViolatingPair(const CArray<double>& alpha, const CArray<double>& gradient,
-	int& i, double& Gmax, int&j, double& Gmin) const
+// j: minimizes the decrease of obj value
+//  (if quadratic coefficient <= 0, replace it with tau)
+//  -y_j*grad(f)_j < -y_i*grad(f)_i, j in I_low(\alpha)
+bool CSMOptimizer::selectWorkingSet( int& outI, int& outJ ) const
 {
-	Gmax = -Inf; Gmin = Inf;
-	i = j = -1;
-
-	for(int t = 0; t < data->GetVectorCount(); t++) {
-		if(data->GetBinaryClass(t) == +1) {
-			if(alpha[t] < getVectorWeight(t)) {
-				if(-gradient[t] >= Gmax) {
-					Gmax = -gradient[t];
-					i = t;
+	double gMax = -Inf;
+	double gMax2 = -Inf;
+	int gMaxIdx = -1;
+	int gMinIdx = -1;
+	double objDiffMin = Inf;
+
+	for( int t = 0; t < activeSize; ++t) {
+		if( y[t] == 1 ) {
+			if( alphaStatus[t] != AS_UpperBound ) {
+				if( -g[t] >= gMax ) {
+					gMax = -g[t];
+					gMaxIdx = t;
 				}
 			}
-			if(alpha[t] > 0) {
-				if(-gradient[t] <= Gmin) {
-					Gmin = -gradient[t];
-					j = t;
-				}
+		} else if( alphaStatus[t] != AS_LowerBound ) {
+			if(g[t] >= gMax) {
+				gMax = g[t];
+				gMaxIdx = t;
 			}
-		} else {
-			if(alpha[t] < getVectorWeight(t)) {
-				if(gradient[t] <= Gmin) {
-					Gmin = gradient[t];
-					j = t;
+		}
+	}
+
+	void ( *updateMinParams )( double, double, const float*, const double*, double, int, double, double,
+		int&, double& );
+	const float* Q_i = NULL;
+	double y_i = 0, QD_i = 0;
+	if( gMaxIdx != -1 ) {
+		Q_i = Q->GetColumn( gMaxIdx, activeSize );
+		y_i = y[gMaxIdx];
+		QD_i = QD[gMaxIdx];
+
+		updateMinParams = []( double gradDiff, double mult, const float* Q_i, const double* QD, double QD_i,
+				int j, double y_i, double Tau, int& gMinIdx, double& objDiffMin ) {
+			if( gradDiff > 0) {
+				double quadCoef = QD_i + QD[j] + mult * y_i * Q_i[j];
+				if( quadCoef <= 0 ) {
+					quadCoef = Tau;
+				}
+				double objDiff = -( gradDiff * gradDiff ) / quadCoef;
+				if( objDiff <= objDiffMin ) {
+					gMinIdx = j;
+					objDiffMin = objDiff;
 				}
 			}
-			if(alpha[t] > 0) {
-				if(gradient[t] >= Gmax) {
-					Gmax = gradient[t];
-					i = t;
+		};
+	} else {
+		updateMinParams = []( double, double, const float*, const double*, double, int, double, double,
```

Comment:
That looks VEEEERY suspicious.

If we're here then we're not gonna update `gMinIdx` and `objDiffMin`. In that case `gMinIdx`  will be equal to `-1` and we will definitely return `true` from this method.

This is why we can omit the cycle below and `return true;` here. And down below we're gonna need to check only the `tolerance` criterion

Reply:
Also, we can rewrite this code in a simpler way
```c++
if( gMaxIdx == -1 ) {
    return true;
}
// Here comes the lambda declaration, the check of the tolerance criterion, etc.
```

Reply:
And that you suggest was my first version of code) But the chain of debugging resulted in this monster... You're right, it needs to be refactored